external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -1497,18 +1483,80 @@ fn exp_inner(context: &mut Context, sp!(eloc, ne_): N::Exp) -> T::Exp {
+                        from_user: true,
+                    },
+                ),
+                TE::UnresolvedError => (er.ty, TE::UnresolvedError),
+                er_ => {
+                    let mut msg = "Invalid 'move'. Expected a variable.".to_owned();
+                    if matches!(er_, TE::Constant(_, _)) {
+                        msg = format!("{msg}. Cannot 'move' constants");
+                    }
+                    context
+                        .env
+                        .add_diag(diag!(TypeSafety::InvalidMoveOp, (loc, msg)));
+                    (context.error_type(eloc), TE::UnresolvedError)
+                }
+            }
+        }
+        NE::ExpDotted(DottedUsage::Copy(loc), sp!(_, N::ExpDotted_::Exp(ner))) => {
+            let er = exp_(context, *ner);
+            let (ty, ecopy) = match er.exp.value {
+                TE::Use(var) => (
+                    er.ty,
+                    TE::Copy {
+                        var,
+                        from_user: true,
+                    },
+                ),
+                er_ @ TE::Constant(_, _) => {
+                    context.env.check_feature(
+                        FeatureGate::PathRework,
+                        context.current_package(),
+                        loc,
+                    );
+                    (er.ty, er_)
+                }
+                TE::UnresolvedError => (er.ty, TE::UnresolvedError),
+                _ => {
+                    let msg = "Invalid 'copy'. Expected a variable or path.".to_owned();
+                    context
+                        .env
+                        .add_diag(diag!(TypeSafety::InvalidCopyOp, (loc, msg)));
+                    (context.error_type(eloc), TE::UnresolvedError)
+                }
+            };
+            if !matches!(ecopy, TE::UnresolvedError) {
+                context.add_ability_constraint(
+                    eloc,
+                    Some(format!(
+                        "Invalid 'copy' of owned value without the '{}' ability",
+                        Ability_::Copy
+                    )),
+                    ty.clone(),
+                    Ability_::Copy,
+                );
+            }
+            (ty, ecopy)
+        }
+
+        NE::ExpDotted(DottedUsage::Borrow(mut_), ndotted) => {
             let (edotted, _) = exp_dotted(context, Some("borrow"), ndotted);
             let eborrow = exp_dotted_to_borrow(context, eloc, mut_, edotted);
             (eborrow.ty, eborrow.exp.value)
         }
 
-        NE::DerefBorrow(ndotted) => {
-            assert!(!matches!(ndotted, sp!(_, N::ExpDotted_::Exp(_))));
-
+        NE::ExpDotted(usage, ndotted) => {
             let (edotted, inner_ty) = exp_dotted(context, Some("dot access"), ndotted);
-            let ederefborrow = exp_dotted_to_owned_value(context, eloc, edotted, inner_ty);
+            let ederefborrow = exp_dotted_to_owned_value(context, usage, eloc, edotted, inner_ty);
```

Comment:
Just thinking out loud: typing does a lot of expansion work that can only be done once type inference has occurred, but may not strictly need to happen in the pass. Have you considered making a new pass between typing and HLIR lowering to do these sorts of expansions?

Things like:
- Expand these dotted lookups
- Expand and/or binary operations (plus possibly break up larger ones)
- Discharging method call syntax

This isn't a change I feel strongly about, but it would help me because I need a place to stick pattern match compilation and there isn't currently a good one: it needs to happen after typing, but doing it during HLIR also seems a bit much.

Reply:
I think we had discussed offline. I think a lot of this makes sense to me though it would definitely be outside of the scope of this PR

Reply:
Agreed about it being outside of the scope of this PR, just wanted to put the thought out into the universe. :)